server/CONTRIBUTING.md

```diff
@@ -112,10 +111,251 @@ pytest --hge-urls http://127.0.0.1:8080 --pg-urls <database_url> -vv
+  | Leaf
+```
+
+```haskell
+data HttpException
+    = InvalidStatusCode Int
+    | MissingContentHeader
+```
+
+Format records as follows:
+
+```haskell
+data Person
+  = Person
+  { firstName :: !String  -- ^ First name
+  , lastName  :: !String  -- ^ Last name
+  , age       :: !Int     -- ^ Age
+  } deriving (Eq, Show)
+```
+
+#### List Declarations
+
+Align the elements in the list.  Example:
+
+```haskell
+exceptions =
+  [ InvalidStatusCode
+  , MissingContentHeader
+  , InternalServerError
+  ]
+```
+
+Optionally, you can skip the first newline.  Use your judgement.
+
+```haskell
+directions = [ North
+             , East
+             , South
+             , West
+             ]
+```
+
+#### Export Lists
+
+Format export lists as follows:
+
+```haskell
+module Data.Set
+  (
+    -- * The @Set@ type
+    Set
+  , empty
+  , singleton
+
+    -- * Querying
+  , member
+  ) where
+```
+
+### Imports
+
+Imports should be grouped in the following order:
+
+1. standard library imports / related third party imports
+2. qualified imports of standard library / third party imports
+3. local application/library specific imports
+4. qualified imports of local application/library specific imports
+
+Put a blank line between each group of imports. The imports in each group should
+be sorted alphabetically, by module name (this is done by stylish-haskell
+automatically).
+
+Always use explicit import lists or `qualified` imports for standard and third
+party libraries. This makes the code more robust against changes in these
+libraries. Exception: The Prelude.
+
+### Naming
+
+Use camel case (e.g. `functionName`) when naming functions and upper camel case
+(e.g. `DataType`) when naming data types.
+
+For readability reasons, don't capitalize all letters when using an
+abbreviation. For example, write `HttpServer` instead of `HTTPServer`.
+Exception: Two letter abbreviations, e.g. `IO`.
+
+#### Modules
+
+Use singular when naming modules e.g. use `Data.Map` and
+`Data.ByteString.Internal` instead of `Data.Maps` and
+`Data.ByteString.Internals`.
+
+### Standard library usage
+
+#### Don't use `unless`
+
+Use `when` combined with `not`. Usage of `unless` can be confusing.
+
+### Data types
+
+#### Wrap primitives in newtype (or sum types)
+
+It is usually a good idea to introduce a custom data type (sum types or
+`newtype`) instead of using a primitive type (like `Int`, `String`, `Set`
+`Text`, etc.). This prevents confusion and mistakes of using the wrong types in
+wrong places.
+
+#### Use sum types instead `Bool` where it makes sense
+
+To represent two different states, use sum types rather than a `Bool`.
+Specially, in the case of data types which represents enabled/disabled state.
+The problem with `Bool` is, it is not clear what the default is.
+
+Example:
+
+```haskell
+data ConsoleEnabled
+  = CEDisabled
+  | CEEnabled
+```
+
+### Dealing with laziness
+
+By default, use strict data types and lazy functions.
+
+#### Data types
```

Comment:
I think this is good advice, but why not actually enforce this by configuring `StrictData` project-wide (along with `-funbox-strict-fields`)? I've never worked on a project that did that, but have often thought it might be a good idea. Obviously a little care would need to be taken porting the project to this initially. Thoughts?

Reply:
I have never done this as well.

Reply:
https://github.com/hasura/graphql-engine/issues/3941